File: backend/app/schemas/review.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, ConfigDict


class PlotTaskStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    PARTIAL_SUCCESS = "partial_success"
    FAILED = "failed"
    PERMANENTLY_FAILED = "permanently_failed"
    SUPERSEDED = "superseded"


class MissingField(BaseModel):
    variable_name: str
    level_type: str | None = None
    level_value: int | None = None
    lead_hour: int | None = None
    reason: Literal["file_not_found", "read_error", "dimension_mismatch"]


class ReviewGenerateRequest(BaseModel):
    window_id: str
    version_id: str
    template_id: str


class ReviewGenerateResponse(BaseModel):
    review_id: str
    plot_status: str
    message: str


class ReviewExportRequest(BaseModel):
    review_id: str


class ReviewProductDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    review_id: str
    window_id: str
    version_id: str
    template_id: str
    image_path: str | None = None
    plot_config_path: str | None = None
    plot_input_manifest_path: str | None = None
    plot_code_version: str | None = None
    plot_status: str
    attempt: int
    max_retries: int
    locked_by: str | None = None
    locked_at: datetime | None = None
    next_retry_at: datetime | None = None
    plot_started_at: datetime | None = None
    plot_finished_at: datetime | None = None
    total_panels: int | None = None
    success_panels: int | None = None
    skipped_panels: int | None = None
    missing_fields_json: str | None = None
    error_log_path: str | None = None
    created_at: datetime


class ReviewProductListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    review_id: str
    window_id: str
    version_id: str
    template_id: str
    plot_status: str
    image_path: str | None = None
    total_panels: int | None = None
    success_panels: int | None = None
    skipped_panels: int | None = None
    created_at: datetime


class ReviewProductListResponse(BaseModel):
    items: list[ReviewProductListItem]
    total: int
    page: int
    page_size: int


class ReviewProductVersionItem(ReviewProductListItem):
    version_no: int | None = None
    version_status: str | None = None
    version_created_by: str | None = None
    version_created_at: datetime | None = None

File: frontend/src/api/review.ts
import http from '@/api/http'

export interface ApiResponse<T> {
  code: string
  message: string
  data: T
  trace_id: string
}

export type PlotStatus =
  | 'pending'
  | 'running'
  | 'success'
  | 'partial_success'
  | 'failed'
  | 'permanently_failed'
  | 'superseded'
  | string

export interface ReviewGenerateParams {
  window_id: string
  version_id: string
  template_id: string
}

export interface ReviewGenerateResponse {
  review_id: string
  plot_status: string
  message: string
}

export interface MissingField {
  variable_name?: string
  name?: string
  level_type: string | null
  level_value: number | null
  lead_hour: number | null
  reason: 'file_not_found' | 'read_error' | 'dimension_mismatch' | string
}

export interface ReviewProductDetail {
  review_id: string
  window_id: string
  version_id: string
  template_id: string
  image_path: string | null
  plot_config_path?: string | null
  plot_input_manifest_path?: string | null
  plot_code_version?: string | null
  plot_status: PlotStatus
  attempt?: number
  max_retries?: number
  locked_by?: string | null
  locked_at?: string | null
  next_retry_at?: string | null
  plot_started_at?: string | null
  plot_finished_at?: string | null
  total_panels: number | null
  success_panels: number | null
  skipped_panels: number | null
  missing_fields_json?: string | MissingField[] | null
  error_log_path?: string | null
  created_at: string
}

export interface ReviewProductListItem {
  review_id: string
  window_id: string
  version_id: string
  template_id: string
  plot_status: PlotStatus
  image_path: string | null
  total_panels: number | null
  success_panels: number | null
  skipped_panels: number | null
  created_at: string
}

export interface ReviewProductVersionItem extends ReviewProductListItem {
  version_no?: number | null
  version_status?: string | null
  version_created_by?: string | null
  version_created_at?: string | null
}

export interface ReviewListParams {
  case_id?: string
  window_id?: string
  plot_status?: string
  page?: number
  page_size?: number
}

interface ReviewListResponse {
  items: ReviewProductListItem[]
  total: number
  page: number
  page_size: number
}

export async function generateReview(
  params: ReviewGenerateParams,
): Promise<ApiResponse<ReviewGenerateResponse>> {
  const { data } = await http.post<ApiResponse<ReviewGenerateResponse>>('/review/generate', params)
  return data
}

export async function getPlotTaskStatus(
  reviewId: string,
): Promise<ApiResponse<ReviewProductDetail>> {
  const { data } = await http.get<ApiResponse<ReviewProductDetail>>(`/tasks/plot/${reviewId}`)
  return data
}

export async function getReviewDetail(
  reviewId: string,
): Promise<ApiResponse<ReviewProductDetail>> {
  const { data } = await http.get<ApiResponse<ReviewProductDetail>>(`/review/${reviewId}`)
  return data
}

export async function getReviews(
  params: ReviewListParams = {},
): Promise<ApiResponse<ReviewProductListItem[]>> {
  const { data } = await http.get<ApiResponse<ReviewListResponse | ReviewProductListItem[]>>(
    '/reviews',
    { params },
  )
  const list = Array.isArray(data.data) ? data.data : data.data.items

  return {
    ...data,
    data: list,
  }
}

export async function getReviewsByCase(
  caseId: string,
): Promise<ApiResponse<ReviewProductListItem[]>> {
  const { data } = await http.get<ApiResponse<ReviewProductListItem[]>>(`/review/case/${caseId}`)
  return data
}

export async function getReviewsByWindow(
  windowId: string,
): Promise<ApiResponse<ReviewProductVersionItem[]>> {
  const { data } = await http.get<ApiResponse<ReviewProductVersionItem[]>>(
    `/review/window/${windowId}/versions`,
  )
  return data
}

export async function exportReview(reviewId: string): Promise<void> {
  const { data } = await http.post<Blob>(
    '/review/export',
    { review_id: reviewId },
    { responseType: 'blob' },
  )
  const url = URL.createObjectURL(data)
  const link = document.createElement('a')
  link.href = url
  link.download = `review_package_${reviewId}.zip`
  document.body.appendChild(link)
  link.click()
  link.remove()
  URL.revokeObjectURL(url)
}
